ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryExecutor.java

```diff
@@ -647,4 +671,107 @@ private static void updateListProperty(
     valueList.add(value);
     properties.put(key, valueList);
   }
+
+  private static class TransientQuerySinkProcessor implements
+      Processor<Object, GenericRow, Void, Void> {
+
+    private final TransientQueryQueue queue;
+    private final Optional<ImmutableMap<TopicPartition, Long>> endOffsets;
+    private final ConcurrentHashSet<TopicPartition> donePartitions;
+    private ProcessorContext<Void, Void> context;
+
+    TransientQuerySinkProcessor(
+        final TransientQueryQueue queue,
+        final Optional<ImmutableMap<TopicPartition, Long>> endOffsets,
+        final ConcurrentHashSet<TopicPartition> donePartitions) {
+
+      this.queue = queue;
+      this.endOffsets = endOffsets;
+      this.donePartitions = donePartitions;
+    }
+
+    @Override
+    public void init(final ProcessorContext<Void, Void> context) {
+      Processor.super.init(context);
+      this.context = context;
+      if (endOffsets.isPresent()) {
+        // registering a punctuation here is the easiest way to check
+        // if we're complete even if we don't see all records.
+        // The in-line check in process() is faster,
+        // but some records may be filtered out before reaching us.
+        context.schedule(
+            Duration.ofMillis(100),
+            PunctuationType.WALL_CLOCK_TIME,
+            timestamp -> checkCompletion()
```

Comment:
We check for completion both periodically and when reading records.  For my education, why is the periodic one needed as well?

Reply:
You're right; I should add a comment :) 

The `process` call is only triggered when processing a record. If an upstream processor filters records out, then this processor (which is right at the end of the topology) might never get triggered and hence might never complete. I figured that out the hard way :/